# Update Null Values

In [51]:
import pandas as pd
import numpy as np

## Load Data

In [52]:
data = pd.read_csv('final_project_mod.csv', dtype = {'Continent': 'category', 'Month':'category', 'DayOfWeek':'category'})

### Explicitly Define Order Category for Month and DayOfWeek

In [53]:
from pandas.api.types import CategoricalDtype

In [54]:
# The fist step in ordering a category feature is to define a custom order by way of CategoricalDtype
month_ordered_categories = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']

# Next, the attribute needs to be explicitly re_ordered:
#data.Month.astype('category', ordered=True,  categories=month_ordered_categories)
data['Month'] = data['Month'].astype(CategoricalDtype(categories=month_ordered_categories))

# Repeat steps 
#dayofweek_ordered_categories = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri']
#dayofweek_ordered_categories_type = CategoricalDtype(categories = dayofweek_ordered_categories, ordered=True)
#data['DayOfWeek'] = data['DayOfWeek'].cat.reorder_categories(dayofweek_ordered_categories, ordered=True)


In [55]:
data.head()

x0        x1         x2        x3        x4         x5         x6  \
0 -0.166563 -3.961588   4.621113  2.481908 -1.800135   0.804684   6.718751   
1 -0.149894 -0.585676  27.839856  4.152333  6.426802  -2.426943  40.477058   
2 -0.321707 -1.429819  12.251561  6.586874 -5.304647 -11.311090  17.812850   
3 -0.245594  5.076677 -24.149632  3.637307  6.505811   2.290224 -35.111751   
4 -0.273366  0.306326 -11.352593  1.676758  2.928441  -0.616824 -16.505817   

          x7        x8        x9  ...        x41       x42       x43  \
0 -14.789997 -1.040673 -4.204950  ...  -1.497117  5.414063 -2.325655   
1  -6.725709  0.896421  0.330165  ...  36.292790  4.490915  0.762561   
2  11.060572  5.325880 -2.632984  ...  -0.368491  9.088864 -0.689886   
3 -18.913592 -0.337041 -5.568076  ...  15.691546 -7.467775  2.940789   
4  27.532281  1.199715 -4.309105  ... -13.911297 -5.229937  1.783928   

        x44       x45        x46       x47       x48        x49  y  
0  1.674827 -0.264332  60.781427 -7.689696  0.151589  -8.040166  0  
1  6.526662  1.007927  15.805696 -4.896678 -0.320283  16.719974  0  
2 -2.731118  0.754200  30.856417 -7.428573 -2.090804  -7.869421  0  
3 -6.424112  0.419776 -72.424569  5.361375  1.806070  -7.670847  0  
4  3.957801 -0.096988 -14.085435 -0.208351 -0.894942  15.724742  1  

[5 rows x 51 columns]

In [56]:
#data.dtypes
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 160000 entries, 0 to 159999
Data columns (total 51 columns):
x0           159974 non-null float64
x1           159975 non-null float64
x2           159962 non-null float64
x3           159963 non-null float64
x4           159974 non-null float64
x5           159963 non-null float64
x6           159974 non-null float64
x7           159973 non-null float64
x8           159979 non-null float64
x9           159970 non-null float64
x10          159957 non-null float64
x11          159970 non-null float64
x12          159964 non-null float64
x13          159969 non-null float64
x14          159966 non-null float64
x15          159965 non-null float64
x16          159974 non-null float64
x17          159973 non-null float64
x18          159960 non-null float64
x19          159965 non-null float64
x20          159962 non-null float64
x21          159971 non-null float64
x22          159973 non-null float64
x23          159953 non-null float64
C

In [57]:
data['Month'].dtypes

CategoricalDtype(categories=['Apr', 'Aug', 'Dec', 'Feb', 'Jan', 'Jul', 'Jun', 'Mar',
                  'May', 'Nov', 'Oct', 'Sep'],
                 ordered=False)

## Locate Null Values: Column by Row - Populate Dict

Column by row where value is null

In [58]:
data_notnull = data.notnull()
nullValues = {}
for column in data:
    nullValues.update({column: data_notnull[data_notnull[column]==False].index.to_list()})

Display the ten index's (Rows) where there is missing values (NaN) in attribute x0.

In [59]:
nullValues['x0'][0:10]

[1282, 2290, 7491, 15484, 22497, 23192, 28954, 35999, 53729, 72318]

Proving there are NaN's within the first ten locations within the x0 column

In [60]:
# data.loc[ROW, COLUMN]
data.loc[nullValues['x0'][0:10] ,'x0']

1282    NaN
2290    NaN
7491    NaN
15484   NaN
22497   NaN
23192   NaN
28954   NaN
35999   NaN
53729   NaN
72318   NaN
Name: x0, dtype: float64

## Distinct Rows With Missing Value

In [61]:
unique_index = []
for key, value in nullValues.items():
    for i in value:
        if i not in unique_index: 
            unique_index.append(i)
unique_index.sort()

In [62]:
print('There are', len(unique_index), 'DISTINCT rows with a NaN in at least one of the attributes.')

There are 1608 DISTINCT rows with a NaN in at least one of the attributes.


## Replace NaN With arithmetic averages (Mean, Median, Mode, etc)

In [63]:
# 
for column in data.columns:
    if (data[column].dtype.name != "category"):
        print(column, data[column].mean())

x0 -0.001027911607636433
x1 0.0013583132941498157
x2 -1.150144637185296
x3 -0.02463729930505865
x4 -0.0005489999661715512
x5 0.013581552327766686
x6 -1.6706701553089789
x7 -7.69279492606893
x8 -0.030540285688330525
x9 0.005462093779336017
x10 0.0022534547682891513
x11 0.030232406883283595
x12 -1.334402055040233
x13 0.007669015129560133
x14 0.008103587049699245
x15 0.0012147158213739453
x16 0.00622314434251844
x17 0.012039895626624008
x18 0.012693959952312336
x19 0.02455542795199547
x20 0.29907406094445654
x21 -0.02913741802162531
x22 0.00840002026083034
x23 0.7220283102008835
x25 -0.0008063103652548311
x26 -0.0010657021192542805
x27 -0.004158613376841231
x28 0.031542770877710465
x31 -0.005945088975282879
x32 -1.0877107439566393e-05
x33 -0.006567455207638439
x34 -0.00042594897562077015
x35 0.0009358660086478089
x36 0.006452519002927416
Money 0.3776284090838059
x38 6.059129567756921
x39 0.0042525128492550785
x40 -2.3165258156321022
x41 6.701075903605524
x42 -1.8338195176344874
x43 -0.002

What is the mode for column x0

In [12]:
data['x49'].mean()

-0.6742238327007368

Return to nan within the first location (row 1282) within the x0 column to see that it has been replace with the mode

In [12]:
data_copy.loc[nullValues['x0'][0] ,'x0']

-1.5926345910622286